src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,184 @@
+import { z } from "zod";
+import fetch from "node-fetch";
+import { Message, CommandArg, word } from "../types";
+
+const USAGE: string =
+  "Usage: `?rankup [<username> --target=(username|rank) --language=(py|js|...) --mode={least|each|spread} --limit=(1-8[kyu])]`";
+
+/*
+`?rankup` calculates the number of katas of each rank required for a user to rank up.
+This command is just for fun, however should increase competition between users,
+and (hopefully) increase motivation.
+
+If no username is passed, server nickname of user is taken instead.
+
+Options:
+  <username>   The username to check. Defaults to users discord server nickname
+  --target=    The rank to reach, or a username of a user to overtake. Defaults to users next rank
+  --language=  The language to inspect. Defaults to "overall"
+  --mode=      Calculation mode.
+                  Least favours 1kyus for least overall katas.
+                  Each finds number of katas per rank individually
+                  Spread (default) prioritises high katas, but also distributes points to lower ones
+  --limit=     The max rank to use in calculations ([1-8]kyu)
+*/
+
+enum Mode {
+  Least = "least",
+  Each = "each",
+  Spread = "spread", // default
+}
+const DEFAULTMODE: Mode = Mode.Spread;
+const SPREADWEIGHT: number = 0.6; // used for Spread mode
+
+// Convert synonyms into the names used by CW API
+const LANGSYNONYMS: { [lang: string]: string } = {
+  py: "python",
+  js: "javascript",
+  hs: "haskell",
+  brainfuck: "bf",
+  cs: "coffeescript",
+  "c#": "csharp",
+  ts: "typescript",
+  "c++": "cpp",
+  "f#": "fsharp",
+  objectivec: "objc",
+  "obj-c": "objc",
+  "objective-c": "objc",
+  coldfusionmarkuplanguage: "cfml",
+  coldfusion: "cfml",
+};
+
+// https://docs.codewars.com/gamification/ranks
+// 3-8 dan added speculatively, based on trend
+const RANKTHRESHOLDS: { [rank: string]: number } = {
+  "7kyu": 20,
+  "6kyu": 76,
+  "5kyu": 229,
+  "4kyu": 643,
+  "3kyu": 1768,
+  "2kyu": 4829,
+  "1kyu": 13147,
+  "1dan": 35759,
+  "2dan": 97225,
+  "3dan": 264324,
+  "4dan": 718593,
+  "5dan": 1953551,
+  "6dan": 5310860,
+  "7dan": 14437910,
+  "8dan": 39250354,
+};
+
+const RANKPOINTS: { [rank: string]: number } = {
+  "8kyu": 2,
+  "7kyu": 3,
+  "6kyu": 8,
+  "5kyu": 21,
+  "4kyu": 55,
+  "3kyu": 149,
+  "2kyu": 404,
+  "1kyu": 1097,
+};
+
+const capitalise = (s: string): string =>
+  s.length ? s[0].toUpperCase() + s.slice(1).toLowerCase() : "";
+
+const getLang = (lang: string): string => LANGSYNONYMS[lang.toLowerCase()] || lang.toLowerCase();
+
+// Retrieve a users language score, or overall score if language is undefined
+const getUserLangScore = async (user: string, lang?: string): Promise<number | undefined> => {
+  return await fetch("https://www.codewars.com/api/v1/users/" + user)
+    .then((response: any): any => (response.status === 404 ? { success: false } : response.json()))
```

Comment:
This sucks, and I'm pretty sure it's a bug. But CW API v1 has been like this as long as I can remember, so fixing this can break some code.

Reply:
Wait, the API does respond with status code `404`. I just misread this.